main.go

```diff
@@ -154,3 +227,179 @@ func gatekeeperNamespace(platformInfo platform.PlatformInfo) (string, error) {
+		Name:      "gatekeeper",
+	}, gatekeeper)
+	if err != nil {
+		setupLog.Error(err, "Getting gatekeeper resource has error")
+
+		return false
+	}
+
+	audit := gatekeeper.Spec.Audit
+	if audit == nil {
+		return false
+	}
+
+	auditFromCache := audit.AuditFromCache
+	if auditFromCache == nil {
+		return false
+	}
+
+	auditFromCacheValue := *auditFromCache
+	if err != nil || auditFromCacheValue != operatorv1alpha1.AuditFromCacheAutomatic {
+		setupLog.V(1).Info("AuditFromCache is not set")
+
+		return false
+	}
+
+	return true
+}
+
+// Check constraintpodstatuses.status.gatekeeper.sh crd status has "NamesAccepted" condition.
+func (c *ContraintStatusInstaller) isApiAvailable(result watch.Event) bool {
+	obj := result.Object.DeepCopyObject()
+	crd, ok := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if ok != nil {
+		return false
+	}
+
+	conditions, _, err := unstructured.NestedSlice(crd, "status", "conditions")
+	if err != nil {
+		return false
+	}
+
+	for _, condition := range conditions {
+		conditionObj, ok := condition.(map[string]interface{})
+		if !ok {
+			return false
+		}
+
+		conditionType := conditionObj["type"].(string)
+		if conditionType == string(v1.NamesAccepted) {
+			return true
+		}
+	}
+
+	setupLog.V(1).Info("Constraintpodstatuses crd is not ready, Wait for crd status update")
+
+	return false
+}
+
+// Watch the constraintpodstatuses resource in this function. When adding or Modified events comes,
+// This function adds constraintpodstatus function. Modified events are expected as status changes
+// When deleting event comes, close the function using context
+func (c *ContraintStatusInstaller) watchGatekeeperInstall(mainCtx context.Context, dynamicClient *dynamic.DynamicClient,
+	cfg *rest.Config, managerOptions manager.Options, namespace string,
+) {
+	contraintStatusCtx := context.Background()
+	ctx, ctxCancel := context.WithCancel(contraintStatusCtx)
+
+	fieldSelector := "metadata.name=constraintpodstatuses.status.gatekeeper.sh"
+	timeout := int64(30)
+	crdGVR := schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1",
+		Resource: "customresourcedefinitions",
+	}
+
+	watchFunc := func(options metav1.ListOptions) (watch.Interface, error) {
+		return dynamicClient.Resource(crdGVR).Watch(mainCtx,
+			metav1.ListOptions{FieldSelector: fieldSelector, TimeoutSeconds: &timeout})
+	}
+
+	watcher, _ := toolsWatch.NewRetryWatcher(crdGVR.Version, &cache.ListWatch{WatchFunc: watchFunc})
+
+	for {
+		select {
+		case result := <-watcher.ResultChan():
+			switch result.Type { //nolint:exhaustive
+			case watch.Modified, watch.Added:
+				if !c.getIsInstalled() {
+					isAutomaticSet := c.getAutomaticSet(mainCtx)
+					if !isAutomaticSet {
+						break
+					}
+					// Check constraintpodstatuses crd available
+					if !c.isApiAvailable(result) {
+						break
+					}
+					c.mu.Lock()
+					contraintStatusCtx = context.Background()
+					ctx, ctxCancel = context.WithCancel(contraintStatusCtx)
```

Comment:
I nee to create new context because you asked me to handle the delete event so when the delete event happens, I had to close the context. Using mainCtx close gatekeeper controller too